### ISSUE TITLE
2 different overloads are incorrectly considered equivalent

### ISSUE THREAD

**Issue by Aran-Fey** [microsoft/pyright#9605]
Code:

```python
class MUL[X, Y]: pass
class DIV[X, Y]: pass

@overload
def divide[X,Y](a: MUL[X, Y], b: X) -> Y: ...

@overload
def divide[X,Y](a: MUL[X, Y], b: Y) -> X: ...
```

Here, pyright complains that:

> Overload 2 for "divide" will never be used because its parameters overlap overload 1  (reportOverlappingOverload)

I don't see how `a: MUL[X, Y], b: X` and `a: MUL[X, Y], b: Y` overlap.

Type inference is consequently also wrong:

```python
int_times_str: MUL[int, str] = ...  # type: ignore
reveal_type(divide(int_times_str, 3))  # Correct: str
reveal_type(divide(int_times_str, 'hi'))  # Also str?!
```

**VS Code extension or command-line**

Command line, version 1.1.391

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so I don't consider this a bug.

There is no list of arguments that will fail to use the first overload. You can see this by deleting the second overload and making the first overload a simple (non-overloaded) function. 

Not surprisingly, mypy emits the same errors as pyright in this example.

Perhaps your intent was for the type parameters in `MUL` to be invariant? They are currently covariant, which allows for this behavior. If you were to modify `MUL` in a way that forces `X` to be invariant, then you'll see that pyright (and mypy) no longer mark the second overload as fully overlapping.

```python
class MUL[X, Y]:
    x: X  # This results in X being inferred as invariant rather than covariant
```

**Comment by Aran-Fey** [comment]
I don't understand. How can `divide(int_times_str, 'hi')` possibly result in a `str`? `int_times_str` is explicitly annotated as being of type `MUL[int, str]`, so `X` must be `int` and `Y` must be `str` . That means the overloads are `(a: MUL[int, str], b: int) -> str` (doesn't match because `'hi'` is clearly not an `int`) and `(a: MUL[int, str], b: str) -> int` (matches, and yet the output *isn't* `int`).

**Comment by erictraut** [comment]
Consider this variation on your code sample:

```python
class A[X, Y]:  # X and Y are invariant
    x: X
    y: Y

class B[X, Y]:  # X and Y are covariant
    pass

def func_a[X, Y](a: A[X, Y], b: X) -> tuple[X, Y]: ...
def func_b[X, Y](a: B[X, Y], b: X) -> tuple[X, Y]: ...

def func(val_a: A[int, str], val_b: B[int, str]):
    reveal_type(func_a(val_a, ""))  # Error: no solution
    reveal_type(func_b(val_b, ""))  # tuple[int | str, str]
```

If the type variables `X` and `Y` are invariant (as they are with class `A`), the call to `func_a` requires the constraint solver to solve the following constraints:

```
X == int
Y == str
X >= str
```

There is no solution to this set of constraints, so an error results.

Now consider `func_b`. The constraints are now:

```
X >= int
Y >= str
X >= str
```

These constraints allow for a valid solution. The simplest is `X == int | str` and `Y = str`. You can see this is the solution pyright (and mypy) arrive at. (I replaced the return type with a tuple to make it easy to see the final solution for both `X` and `Y`.)

**Comment by erictraut** [comment]
> It ignores the specified type and infers a Union instead

It's not "ignoring" the type. The type checker is taking the type into consideration when it constructs the list of constraints it needs to solve. When a constraint is invariant, the constraint solver has less freedom. When a constraint is covariant, the constraint solver can widen the type (and similarly narrow the type if the constraint is contravariant) when looking for a type that satisfies all of the constraints.

Here's a pro tip to help you with variance inference. If you are using the language server version of pyright (versus the command-line version), you can hover over the type parameter in the class definition, and the hover text will display the inferred variance in parentheses.

![Image](https://github.com/user-attachments/assets/b8965ca4-c33a-4246-bb10-2b10ff2b4fdd)

> I still can't quite wrap my head around this

Yeah, it can be confusing. The notion of constraint solving took me a while to wrap my head around too. It's not well documented anywhere. I'd like to eventually add a section to the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) that covers this topic in more detail.
